[trainer] Support parsing test results from SwiftTesting

AliSoftware opened the pull request:
This PR allows `trainer` (and by extension `scan`/`run_tests`) to support test results from [Swift Testing](https://developer.apple.com/documentation/Testing)—by parsing the `.xcresult` bundle using the new `xcresulttool` API

### ☑️  Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary.
- [x] I've added or updated relevant unit tests.

### ❔  Motivation and Context

In Xcode 16b3, the `xcresulttool` tool—that ships with Xcode Command Line Tools to parse and extract data from `.xcresult` bundles—had its CLI API updated.

So far we've relied on adding the `--legacy` flag to our `xcresulttool` invocations to be able to still execute old CLI and still use the old way of calling the tool, using `xcresulttool get <object> --format json` / `xcresulttool get <object> --id <id>`. But:
 - As the name of the flag suggests, this way of calling the tool is legacy, and might become unsupported in future versions
 - That way of calling the tool seems to not support extracting information from reports from tests written using Swift Testing, and only supports tests written in XCTest.

The version of `xcresulttool` that ships with Xcode 16+ now expects us to call the CLI with `xcresulttool get test-results test …`, and generates a new JSON format (whose JSONSchema is described in `xcresulttool get test-results test --help`) that is easier to parse but more importantly also supports Swift Testing test results.

### ✍️  Description

Instead of trying to modify the implementation of `xcresult.rb` to adapt to the new command and JSON format—which would make the code difficult to read if we wanted to keep backwards compatibility in the same codebase—and because the new JSON format is way better structured and doesn't need the complexity of what the old XCResult parser implementation needed, I've decided to instead avoid piling up on the legacy implementation and implement the parsing of new JSON format completely separately.

I have thus:
 - Extracted the methods dedicated to parsing the legacy `.xcresult` format as well as the old `.plist` format in dedicated classes and files, to separate that legacy code clearly.
 - Renamed the class dedicated to parsing the legacy `.xcresult` format to `LegacyXCResult`
 - Created a new `XCResult` class instead, dedicated to parsing the new format
 - Made `TestParser` call the `parse_xcresult` method of either `LegacyXCResult` or `XCResult` class depending on the version of `xcresulttool` being used/installed (with a possibility to force the legacy mode if needed, to avoid breaking things accidentally if we missed some edge cases in the new implementation

The new implementation of `XCResult` parsing is also now split into nested classes under `xcresult/*.rb`, namely `TestPlan`, `TestSuite`, `TestCase`, and `Repetitions`, which make semantic sense and are easier to reason about when parsing the `xcresult` JSON representation.

The generation of the output JUnit file is also directly handled in the model objects (`to_xml` methods) rather than using ERB (via `JunitGenerator`) like the old implementation was using.

### 🧪  Testing Steps

I've added a couple of Unit Tests to `test_parser_spec.rb`, that use an `.xcresult` file that I generated from a dummy `xcproject`[^1], which contained a mix of Unit and UITests, implemented with a mix of XCTest and SwiftTesting, and using features like test skipping, arguments, and retries.

Hopefully this should cover a good set of features that those testing frameworks offer.

That being said, I'd love if more people would test this implementation against some `.xcresult` files of their own Xcode projects, to validate the generated JUnit `.xml` looks right and contains all necessary info (failure messages, etc) and so that it covers a whole range of possibilities (e.g. tests runs with or without retries enabled, custom traits, etc…)

### 👁️  Suggestions on how to review this PR

This PR's diff is huge, but this is mostly because:
 - A lot of code was just moved (as opposed to modified/added/deleted)
 - I included large fixtures (`.junit` expectation fixtures which are XMLs of >1K lines each)
 - I commited the dummy Xcode project I used for generating the `.xcresult` fixture, even if that Xcode project is not directly referenced by any of the rspec tests, but so that it'd be easy to add more tests in the future in we wanted to regenerate the `.xcresult` to cover more edge cases
 - ~~I also had to commit the new `.xcresult` fixture… which is in practice a bundle containing a lot of small (mostly-binary) files. So that shows up as a lot of files in this PR's diff, while those don't need to be manually reviewed at all when reviewing that PR.~~ [EDIT] I ended up only commiting the `.xcresult/database.sqlite3` file as I discovered that it was enough for `xcresulttool` to operate.

I'd thus suggest the reviewers to instead focus on:
 - Changes in `trainer/lib/*` — keeping in mind that some of those changes are just existing code for the legacy parsers having been moved to dedicated files 
 - Unit tests in `trainer/spec/*_spec.rb`
 - For `trainer/spec/fixtures`, only check the `*.junit` expectation fixtures that were added. You can in particular safely ignore:
    - The whole `trainer/spec/fixtures/Xcode16-Mixed-XCTest-SwiftTesting.xcresult/*` folder and its files, since that fixture was auto-generated (it's the bundle that Xcode generated from my dummy project)
    - The intermediate `Xcode16-Mixed-XCTest-SwiftTesting.json` fixture, as that fixture was also auto-generated (by running `xcresulttool` on the added `.xcresult` bundle fixture)
    - The whole `trainer/spec/fixtures/Xcode16-Mixed-XCTest-SwiftTesting-Project/` folder, since this just contains the dummy Xcode project with lots of dummy tests that was used to generate the `.xcresult` fixture. Of course it's ok if you want to review the Xcode project itself to compare how those various ways of writing tests in Swift appear in the JUnit XML fixture files, but it's probably not worth reviewing that in GitHub directly if so, but directly in Xcode instead.

Hopefully those suggestions should help reduce the scope of the review of this PR by a lot

[^1]: Created with Cursor and AI agent, so I could quickly generate a full project with a wide range of all possible API usages and edge cases of XCTest and SwiftTesting features

AliSoftware committed:
Extract parsing methods into dedicated files

Extract parsing of PLIST and Parsing of --legacy xcresulttool output each in their own dedicated file, on preparation for refactoring and introducing new xcresult parsing method for Xcode 16

AliSoftware committed:
Add new XCResult parser + tests

AliSoftware committed:
Clean large attachments and log data from xcresult

To avoid storing useless large data from fixture in git, while it's not used for our test cases

AliSoftware committed:
Fix extracting data when argument child nodes

AliSoftware committed:
Use different <testcase name=…> if using argument

To ensure that tools that parse the generated JUnit would consider those as different test cases, instead of multiple runs of the same test case.

We already have <property name="argument" value="…"> to help us distinguish, but in practice most of the tools that read JUnit XML files will now know about that property or account for it, and only look at the `<testcase classname=…name=…>` attributes to differentiate cases.

AliSoftware committed:
Extract each Trainer::XCResult::* nested class

Into a dedicated file, for easier browsing of the code and smaller files to scroll through.

AliSoftware committed:
Fix JUnitGenerator test

AliSoftware committed:
Move `supports_xcode16_xcresulttool?` to Helper

AliSoftware committed:
Fix `rubocop` violations

AliSoftware committed:
Exclude tests requiring xcresulttool from windows

AliSoftware committed:
Remove unused require

AliSoftware committed:
Skip tests for new XCResult if CI don't support it

AliSoftware committed:
Better report failed xcresulttool calls

AliSoftware committed:
Make tests pass with earlier Xcode

So that even you run the tests on CI with older versions of Xcode (that don't need nor know the `--legacy` flag) would still run and pass those tests

AliSoftware committed:
Improve shell command call

Use Open3 to run the command instead of backticks
 - So that we can capture the error if it happens
 - So that the code is more resilient against special characters
 - So that we avoid the involvement of an intermediate shell (that would have to interpret the string into parameters before passing it to the process, requiring us to `Shellwords.escape` it in the first place… instead of just passing the parameters to the process directly without involving some unnecessary shell escaping + parsing intermediary)

TheMetalCode commented:
Happy to test this with my projects! Thank you! 👏 

I had just started to figure out Swift Testing issues and had started testing with https://github.com/fastlane/fastlane/pull/29470, which addressed issues with test cases nested in inner classes and included some tests specific to Swift Testing, but I saw no way to handle parameterized tests (see https://github.com/fastlane/fastlane/issues/29174) and a more explicit/clean way to detect and remove auto-retries without using the newer, non-legacy `xcresulttool` output and new parsing/modeling code to match. Evidently I missed the presence of this PR until today. 🤦

TheMetalCode commented:
@AliSoftware One thing I'm seeing so far in my testing:

```bash
[16:41:04]: Called from Fastfile at line 154
[16:41:04]: ```
[16:41:04]:     152:      desc "Uses trainer to collect test results after a test run"
[16:41:04]:     153:      lane :collect_test_results do |options|
[16:41:04]:  => 154:        trainer(
[16:41:04]:     155:          path: "#{options[:test_plan]}_test_result_bundle/Test-#{options[:test_plan]}.xcresult",
[16:41:04]:     156:          output_directory: "test_results",
[16:41:04]: ```
[16:41:04]: Illegal character "\u0001" in raw string "audioPlayerReturnedForOnlyValidInput(\u0001)"
```

For a test that looks something like:

```Swift
@Test(.serialized, arguments: (0..<128).map(UnicodeScalar.init).map(Character.init))
func audioPlayerReturnedForOnlyValidInput(asciiCharacter: Character) throws { ... }
```

Hopefully some kind of sanitization we can do on test names? Similar to https://github.com/fastlane/fastlane/blob/alisoftware/trainer/support-swift-testing/trainer/lib/trainer/legacy_xcresult.rb#L180?

AliSoftware commented:
Thanks so much @TheMetalCode for testing this PR against some more concrete tests!

> Hopefully some kind of sanitization we can do on test names? Similar to https://github.com/fastlane/fastlane/blob/alisoftware/trainer/support-swift-testing/trainer/lib/trainer/legacy_xcresult.rb#L180?

Ah, good catch. Yeah, that sanitization seems like a decent plan and easy solution.

I'd like to better understand the root cause of this issue first though rather than blindly applying a workaround that was applied at the time the first `trainer` implementation was done, because maybe we can come up with something better / more resilient? Like, depending on at which level in the pipeline this error happens (during the execution of `xcresulttool` itself? During Ruby's parsing of `xcresulttool`'s JSON stdout to a Hash object? Later in the process when trying to generate the XML?) maybe if we force the use of the right unicode encoding at the right level in the code to avoid the issue altogether? 🤔

AliSoftware commented:
Btw, for refs, below is the dummy Xcode project I used to generate the `trainer/spec/fixtures/Xcode16-Mixed-XCTest-SwiftTesting.xcresult` bundle that I use in my fixtures.

[FastlaneTrainerExample.zip](https://github.com/user-attachments/files/19067975/FastlaneTrainerExample.zip)

Maybe I should commit it in the repo, even if the project itself is not used by the tests, so that it'd be easier to re-generate the `.xcresult` bundle with more tests if we want to add more test cases (like the one you made with tricky unicode characters)?

Speaking of which, feel free to add your test case from above to that dummy Xcode project then regenerate the `.xcresult` and `.json` fixtures with that added test, so we can make sure we cover such edge-case (via a future commit that'd fix this and take care of those kind of uncommon characters)

AliSoftware committed:
Commit dummy project used to generate fixture

Even if it's not directly used/referenced as a fixture by the specs themselves, it can still be useful to have it around to make it easier to add more tests in the project then regenerate the `.xcresult` fixture with more test results cases

AliSoftware committed:
Add test containing tricky characters

Both in its name (emojis) and parameters (ASCII characters 0–128)

AliSoftware committed:
Update .xcresult fixture with new data

AliSoftware committed:
ignore `.sqlite3` files from `.xcresult` fixtures

AliSoftware committed:
Escape special chars when generating JUnit XML

AliSoftware committed:
Regenerate JUnit fixtures to match new xcresult

AliSoftware committed:
Update test: add error message with special chars

To ensure that our sanitizing is not just working for test parameters, but also for error messages and similar strings that could be encountered elsewhere in the generated XML.

This not only include the special, non-printable characters with unexpected ASCII values in the custom error message, but also ensure we include `>` and `<=` characters (which will have to be XML-escaped by REXML to be rendered properly)

This also incidentally validates that a `#expect(…)` with a custom error message is properly taken into account when as a custom error message when parsing the `.xcresult` and generating the JUnit XML

AliSoftware committed:
Fix code comment location

TheMetalCode commented:
@AliSoftware Your latest push definitely resolves the issue I was seeing on my end with the ASCII chars, and looking at your sample Xcode project I think you're covering all the edge cases I could think of. I think it's a great idea to check that into the repo too - adds transparency and helps with future maintenance for sure.

The only issue I see now is that CircleCI throws a non-fatal error when uploading the xmls produced by this trainer patch to test results:

<img width="1368" alt="run_unit_tests-News__667182__-_doximity_iOS-Doximity" src="https://github.com/user-attachments/assets/2f12212a-0caf-4de2-8d5f-1ba7be24c5bf" />

Which is....wrong? `testsuites` and `properties` are totally valid elements to use in a JUnit xml!

Comparing the xmls generated by trainer currently to those generated by this branch, I really don't see a lot of difference in terms of xml structure. For the top-level `testsuites` element I'm seeing `<testsuites tests='1037' failures='0' skipped='1' time='4923.3454'>` vs `<testsuites tests="990" failures="0">`, so more attributes at the top level. And I guess current `trainer` doesn't put `properties` under a `testcase` but again, that's perfectly valid for JUnit xml, see https://github.com/testmoapp/junitxml/blob/main/examples/conventions.xml

The only other difference I notice is single-quotes for attribute values instead of double-quotes, not sure if that's an issue or not.

And it's a non-fatal error - it doesn't fail the CI job, nor does it fail to populate the test results tab:
<img width="905" alt="Cursor_and_run_unit_tests-News__667182__-_doximity_iOS-Doximity" src="https://github.com/user-attachments/assets/a91a1825-2842-421d-81eb-3d26bc7b5c94" />

TheMetalCode commented on trainer/lib/trainer/xcresult/test_plan.rb:
```diff
@@ -0,0 +1,89 @@
+        end
+
+        # Convert to XML string with prologue
+        doc = REXML::Document.new
```
```suggestion
        doc = REXML::Document.new
        doc.context[:attribute_quote] = :quote
```

See https://github.com/fastlane/fastlane/pull/29463#issuecomment-2704511016

Not sure how much single vs double quote matters for attribute values but I believe this would make it double-quotes to be consistent with what [the legacy junit generator](https://github.com/fastlane/fastlane/blob/alisoftware/trainer/support-swift-testing/trainer/lib/trainer/junit_generator.rb) would produce.

TheMetalCode replied:
The usage of `:attribute_quote` doesn't appear to be explicitly documented :(

But you can see it in action here: https://github.com/ruby/rexml/blob/master/test/test_attributes.rb#L181

AliSoftware replied:
Nice one!

But since those single vs double quotes were not the core of the issue CircleCI was having with the XML files (as it was instead the `<properties>` not being expected at the top-level element, as you later discovered), and with the fact that `:attribute_quote` is not officially documented, I'd prefer to avoid using it / relying on it in our implementation after all, and stick to the official approach and defaults of how REXML generates and quotes the attributes. wdyt?

TheMetalCode replied:
Agreed, probably not worth the trouble for something that wasn't actually causing problems.

TheMetalCode commented on trainer/lib/trainer/xcresult/test_plan.rb:
```diff
@@ -0,0 +1,89 @@
+        doc = REXML::Document.new
+        doc << REXML::XMLDecl.new('1.0', 'UTF-8')
+
+        # Add properties for configuration and device
+        properties = Helper.create_xml_element('properties')
+
+        @configurations.each do |config|
+          config_prop = Helper.create_xml_element('property', name: 'Configuration', value: config['configurationName'])
+          properties.add_element(config_prop)
+        end
+
+        @devices.each do |device|
+          device_prop = Helper.create_xml_element('property', name: 'device', value: "#{device.fetch('modelName', 'Unknown Device')} (#{device.fetch('osVersion', 'Unknown OS Version')})")
+          properties.add_element(device_prop)
+        end
+
+        testsuites.add_element(properties) if properties.elements.any?
+
+        doc.add(testsuites)
```
Ok upon further inspection, I think the top-level `testsuites` element is not supposed to have `properties`, so this might actually be the source of the issue I referenced in https://github.com/fastlane/fastlane/pull/29463#issuecomment-2704511016

https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd
https://github.com/testmoapp/junitxml/blob/main/examples/junit-complete.xml

A `testsuite` element can have `properties`, as can a `testcase`, but the top-level `testsuites` element is not expected to have `properties`. 

So when CircleCI says this:
```bash
* invalid testsuites element: properties
* invalid top level end element: testsuites
```

It might be saying that `testsuites` is invalid on account of having `properties`.

It's probably going to seem tedious to put `Configuration` and `device` properties under every `testsuite` but that does appear to be what the JUnit schema expects.

TheMetalCode replied:
On my end, I was able to do a REXML `delete_element("/testsuites/properties")` before the CircleCI `upload_test_results`. I can confirm that having `properties` in the top-level `testsuites` is the only issue, no errors are thrown once those are gone.

AliSoftware replied:
Nice catch and investigation!

I've just pushed [7deb84b](https://github.com/fastlane/fastlane/pull/29463/commits/7deb84b030ed4c5e60871530e09a9573811529b0) which should fix this

TheMetalCode replied:
Confirmed the fix, thank you!

AliSoftware committed:
Only commit database.sqlite3 file for new fixture

The `*.xcresult/database.sqlite3` database file contains the same data as what's serialized in the `.xcresult/Data/*` files, and is just an alternate (and more modern and disk-space-optimized) representation of the same info that the latest `xcresulttool` supports

So we might as well commit just the database instead of the whole Data/* folder and its files, not only to make it take less space on disk, but also to make the git diff and logs easier to deal with.

AliSoftware committed:
Attach <properties> to each <testsuite> instead

As apparently JUnit conventions don't allow the root <testsuites> element to contain <properties> elements, only <testsuite> child elements can

h/t @TheMetalCode for noticing in https://github.com/fastlane/fastlane/pull/29463#discussion_r1983844584

AliSoftware committed:
Fix documentation style violation

`trainer` -> _trainer_

AliSoftware committed:
Fix rubocop violations

AliSoftware committed:
Update doc comments in .gitignore

TheMetalCode commented on trainer/lib/trainer/legacy_xcresult.rb:
```diff
@@ -0,0 +1,586 @@
+module Trainer
+  module LegacyXCResult
+    # Model attributes and relationships taken from running the following command:
+    # xcrun xcresulttool formatDescription
```
```suggestion
    # xcrun xcresulttool formatDescription --legacy
```

Not major since the name of the module is now `LegacyXCResult` so it's possible the `--legacy` flag would be inferred, but just in case anybody was trying to run this command and didn't make the inference.

AliSoftware replied:
Nice catch!

AliSoftware committed:
Update doc comment in legacy_xcresult.rb

Co-authored-by: Jason Hagglund <jhagglund@doximity.com>

mollyIV commented on trainer/lib/trainer/plist_test_summary_parser.rb:
```diff
@@ -0,0 +1,84 @@
+
+      def ensure_file_valid!(raw_json)
+        format_version = raw_json["FormatVersion"]
+        supported_versions = ["1.1", "1.2"]
```
Does it mean that with every new version of Xcode, we need to check if the format has changed and support the new version if necessary?

AliSoftware replied:
Yes and no.

To the best of my understanding, this `.plist` format is legacy format, even older than before `.xcresult` bundles and `xcresulttool` were introduced. This is just existing code that was initially in `xcresult.rb` and that I extracted in its own file just like I did for `legacy_xcresult.rb` so we keep that backwards compatibility, but in practice I'm not sure any modern version of Xcode produces those `.plist` file formats for test summaries anymore, so I don't see a new `FormatVersion` being added ever by now.

mollyIV commented on trainer/lib/trainer/test_parser.rb:
```diff
@@ -102,15 +101,16 @@ def initialize(path, config = {})
 
       if File.directory?(path) && path.end_with?(".xcresult")
-        parse_xcresult(path, output_remove_retry_attempts: config[:output_remove_retry_attempts])
+        klass = XCResult::Helper.supports_xcode16_xcresulttool? && !config[:force_legacy_xcresulttool] ? XCResult::Parser : LegacyXCResult::Parser
```
nitpick: 

```suggestion
        parser = XCResult::Helper.supports_xcode16_xcresulttool? && !config[:force_legacy_xcresulttool] ? XCResult::Parser : LegacyXCResult::Parser
```

AliSoftware replied:
Fixed in 7bd6bcdd9

mollyIV commented on trainer/lib/trainer/xcresult/helper.rb:
```diff
@@ -0,0 +1,53 @@
+        version = match[:version]
+
+        Gem::Version.new(version) >= Gem::Version.new(23_021)
+      rescue
```
Hmm, it seems a `begin` block is missing, isn't it?

AliSoftware replied:
Nope, `begin…end` is only needed when you need to specify a smaller scope for `rescue` than the current scope. Here without an explicit `begin…end`, the `rescue` applies to the whole scope of the `def …` method: there's no need to encapsulate the whole body of `def method_name … end` within a `begin…end` in those cases, as it's implicit already.

See [this rubocop rule](https://docs.rubocop.org/rubocop/cops_style.html#styleredundantbegin) for details and examples.

mollyIV commented on trainer/lib/trainer/xcresult/repetition.rb:
```diff
@@ -0,0 +1,36 @@
+
+module Trainer
+  module XCResult
+    class Repetition
```
What does `repetition` sand for here?

AliSoftware replied:
It's for when you enabled retries in your test plan Xcode (e.g. either retry on failure, or retry N times, etc).

I hesitated into naming it `class Retries` instead of `class Repetition`, but the issue with naming it `Retries` is that it suggests it doesn't count the first run, only the retries if the first run failed. While in reality if the test was run 3 times—i.e. it was run once, failed, was retried once, failed again, and was retried a 2nd time), the JSON output of `xcresulttool` will contain 3 repetitions/retries, not 2.

Besides, [the `nodeType` in the JSON returned by `xcresulttool` is named `Repetition` anyway](https://github.com/fastlane/fastlane/pull/29463/files#diff-cf9f90c4c8e47261aaaa6ccb8091fa2a22d0a0799b5043650cebff8b02c3a395R32), not `Retries`, so I figured I'd match the naming.

AliSoftware replied:
Added some more code comments in 30f32894f to provide some additional doc about this.

mollyIV commented on trainer/lib/trainer/options.rb:
```diff
@@ -50,6 +50,11 @@ def self.available_options
                                      description: "Produces class name and test name identical to xcpretty naming in junit file",
                                      is_string: false,
                                      default_value: false),
+        FastlaneCore::ConfigItem.new(key: :force_legacy_xcresulttool,
```
This new option key is going to help to rollout this implementation with minimized regression risk 👏

mollyIV commented:
Hey Ali,

Thanks a lot for this PR! Supporting SwiftTesting is a very much welcomed addition, and I truly appreciate the effort you put into it.

Also, a special shoutout for the well-written PR description. It was super informative and made reviewing this big PR much smoother. The clear steps you provided helped me a lot!

Great work, and thanks again for contributing! 🚀

mollyIV commented:
> even if that project is not directly used as a fixture by the rspec tests, it's useful to have it around for the future

I agree, we could easily add more test cases in the future there 👍

AliSoftware committed:
Add some mode code comments

Especially to be clearer about the Repetition type

AliSoftware committed:
Rename variable to nicer name

As suggested in https://github.com/fastlane/fastlane/pull/29463/files#r1990796517

AliSoftware commented:
@mollyIV I should have addressed all your suggestions, so ready for a re-review (and hopefully approval) 🙂

TheMetalCode reviewed (approved):
For whatever my approval is worth: I've been continuously running this branch on two iOS projects over the last week or so, just over 10K test cases, about 60-70% of which are Swift Testing. Aside from the ASCII character parsing and junit schema issues that have already been addressed with follow-up commits, I've seen no issues and can happily report that all failing test cases, Swift Testing or otherwise, are being properly reflected in the XML report.

This is a very welcome and deeply appreciated change, thank you! 👏

mollyIV reviewed (commented):
Hi 👋 

I tested the changes on some of our projects - mix of Swift Testing and XCTests, pure XCTests and pure Swift Testing, and all were looking great! I didn't spot any issues. 

Thanks again for the changes, Ali. 

Let's ship it 🫡
